### PR TITLE
JS beautifier

### PR DESCRIPTION
**Pull request by ujjwal96**
Left one used js-beautifier
screen

refs: #2396

### CODE REVIEW

**Comment by MatthewShao** [comment]
Frankly, we're quite concern about the speed of this, the current implementation makes our static viewer generation slow. Does this PR helps? Or do you have any idea in mind that can help to improve the speed? :smiley:

**Comment by ujjwal96** [comment]
@MatthewShao I haven't benchmarked this but the speed difference was quite evident when I tried it against a 48 kb js file. It was faster than the current implementation at least.
Can you try applying the patch locally and benchmark it for the static viewer?

**Comment by MatthewShao** [comment]
Fantastic! I tried it on the same `flows` file about 2MB, which I have been using for the static viewer tests.
The improvement is obvious:
#### before
```
Slow content view: [decoded gzip] JavaScript took 0.2s
Slow content view: [decoded gzip] JavaScript took 0.8s
Slow content view: [decoded gzip] CSS took 0.5s
Slow content view: [decoded gzip] CSS took 2.0s
Slow content view: [decoded gzip] CSS took 0.4s
Slow content view: [decoded gzip] JavaScript took 0.8s
Slow content view: [decoded gzip] JavaScript took 0.8s
Slow content view: [decoded gzip] JavaScript took 0.4s
Slow content view: [decoded gzip] JavaScript took 1.5s
Slow content view: [decoded gzip] JavaScript took 0.6s
Slow content view: [decoded gzip] CSS took 0.1s
Slow content view: [decoded gzip] JavaScript took 0.8s
Slow content view: [decoded gzip] JavaScript took 0.3s
Slow content view: [decoded gzip] JavaScript took 0.1s
Slow content view: [decoded gzip] JavaScript took 0.8s
Slow content view: [decoded gzip] JavaScript took 0.2s
Slow content view: [decoded gzip] JavaScript took 0.9s
Slow content view: [decoded gzip] CSS took 0.2s
Slow content view: [decoded gzip] JavaScript took 1.3s
Slow content view: [decoded gzip] JavaScript took 0.2s
Slow content view: [decoded gzip] JavaScript took 1.0s
```
#### After
```
Slow content view: [decoded gzip] CSS took 0.4s
Slow content view: [decoded gzip] CSS took 2.0s
Slow content view: [decoded gzip] CSS took 0.3s
Slow content view: [decoded gzip] CSS took 0.1s
Slow content view: [decoded gzip] CSS took 0.2s
```
The script will only report a slow content view when it uses more than 0.1s to do the jobs, so from the result we can see that this PR can drop the running time for some tasks from `>1s` to `<0.1s`.

File: mitmproxy/contentviews/javascript.py [hunk 1]
```diff
@@ -12,9 +54,17 @@ class ViewJavaScript(base.View):
...
+            yield [("text", line)]
+
     def __call__(self, data, **metadata):
-        opts = jsbeautifier.default_options()
-        opts.indent_size = 2
         data = data.decode("utf-8", "replace")
-        res = jsbeautifier.beautify(data, opts)
-        return "JavaScript", base.format_text(res)
+        if len(data) < 1024:
+            opts = jsbeautifier.default_options()
```
Comment (mhils) [inline mitmproxy/contentviews/javascript.py]: We should strive to remove jsbeautifier entirely here, that's one dependency less to worry about. :)
(I know we still use it for CSS; should ultimately be removed there as well)

**Comment by ujjwal96** [comment]
For app.js it's taking about 0.4 secconds

File: mitmproxy/contentviews/javascript.py [hunk 2]
```diff
@@ -1,6 +1,82 @@
-import jsbeautifier
+from mitmproxy.utils import strutils
+from mitmproxy.contentviews import base
 
-from . import base
+DELIMITERS = '{};\n'
+SPECIAL_AREAS = (
+    ("'", strutils.NO_ESCAPE + "'"),
+    ('"', strutils.NO_ESCAPE + '"'),
```
Comment (mhils) [inline mitmproxy/contentviews/javascript.py]: We need to add "`" as a (ES6) string delimiter.

**Comment by ujjwal96** [comment]
jsbeautifier take 1.1s for the same file.

**Comment by ujjwal96** [comment]
`Beautifying app.js took 0.026s`

**Comment by mhils** [comment]
After experimenting with this a bit more, the following seems to work great for me with 31ef7f149e4553eb9403634c0eec6de4d0123386:
```python
JS_AREAS = (
    r"/(?:[^\n/]|(?<!\\)(?:\\\\)*\\/)+?/(?=[gimsuy]{0,6}\s*[;,\)\.])",
    r"'" + strutils.MULTILINE_CONTENT_LINE_CONTINUATION + strutils.NO_ESCAPE + "'",
    r'"' + strutils.MULTILINE_CONTENT_LINE_CONTINUATION + strutils.NO_ESCAPE + '"',
    r'`' + strutils.MULTILINE_CONTENT + strutils.NO_ESCAPE + '`',
    r"/\*" + strutils.MULTILINE_CONTENT + "\*/",
    r"//" + strutils.SINGLELINE_CONTENT + "$"
)
```

@ujjwal96: Not sure how that performs though - do you want to give it a try using your benchmark? 😄

**Comment by ujjwal96** [comment]
Takes 0.038s

File: mitmproxy/contentviews/javascript.py [hunk 3]
```diff
@@ -1,6 +1,50 @@
...
+def beautify(data):
+    data = strutils.escape_special_areas(
+        data,
+        SPECIAL_AREAS,
+        DELIMITERS
+    )
+
+    data = re.sub(r"\s*;\s*", ";\n", data)
+    data = re.sub(r"\s*}\s*(?!;)", "\n}\n", data)
+    data = re.sub(r"\s*}\s*(?=;)", "\n}", data)
```
Comment (mhils) [inline mitmproxy/contentviews/javascript.py]: I _think_ we can simplify the two above to
```python
data= re.sub(r"\s*}(;)?\s*",r"}\1\n", data)
```

File: mitmproxy/contentviews/javascript.py [hunk 4]
```diff
@@ -1,6 +1,50 @@
...
+    data = re.sub(r"\s*{\s*(?!};)", " {\n", data)
```
[the remark this answers is not in the document]
Comment (ujjwal96) [reply mitmproxy/contentviews/javascript.py]: It is `};`
It shouldn't match `function(){};` It wasn't working that's a different thing :sweat_smile:

**Pull request merged**